mvrx/src/main/kotlin/com/airbnb/mvrx/CoroutinesStateStore.kt

```diff
@@ -0,0 +1,128 @@
+    private val withStateChannel = Channel<(S) -> Unit>(capacity = Channel.UNLIMITED)
+
+    /**
+     * This combination of BroadcastChannel, mutable state property and channelFlow was arrived
+     * at after multiple attempts. This code needs to:
+     * 1) Multicast
+     * 2) Never fail to deliver an intermediate state to subscribers even if updates are fast or
+     *    some subscribers are slow.
+     * 3) Be able to provide an initial value.
+     *
+     * StateFlow can't be used because it conflates new sets. If you set its value in a tight
+     * loop and listen to changes on a flow, you will miss values. ConflatedBroadcastChannel
+     * alone has the same issue.
+     *
+     * A normal Channel can't be used because it isn't multicast.
+     */
+    private val stateChannel = BroadcastChannel<S>(capacity = Channel.BUFFERED)
+    override var state = initialState
+    override val flow: Flow<S> get() = channelFlow {
+        send(state)
+        stateChannel.consumeEach { send(it) }
+    }.distinctUntilChanged()
+
+    init {
+        setupTriggerFlushQueues(scope)
+        scope.coroutineContext[Job]?.invokeOnCompletion {
+            flushQueuesChannel.close()
+            setStateChannel.close()
+            withStateChannel.close()
+        }
+    }
+
+    /**
+     * Observe [flushQueuesChannel] and flush queues whenever there is a new item.
+     * This no-ops if [MvRxTestOverrides.FORCE_SYNCHRONOUS_STATE_STORES] is set.
+     */
+    private fun setupTriggerFlushQueues(scope: CoroutineScope) {
+        if (MvRxTestOverrides.FORCE_SYNCHRONOUS_STATE_STORES) return
+
+        val executor = Executors.newSingleThreadExecutor()
+        scope.coroutineContext[Job]!!.invokeOnCompletion {
+            executor.shutdownNow()
+        }
+
+        scope.launch(executor.asCoroutineDispatcher()) {
+            flushQueuesChannel.consumeEach {
+                flushQueues()
+            }
+        }
+    }
+
+    /**
+     * Flush the setState and withState queues.
+     * All pending setState reducers will be run prior to every single withState lambda.
+     * This ensures that situations such as the following will work correctly:
+     *
+     * Sitaution 1
+     *
+     * setState { ... }
+     * withState { ... }
+     *
+     * Sitaution 2
+     *
+     * withState {
+     *     setState { ... }
+     *     withState { ... }
+     * }
+     */
+    private fun flushQueues() {
+        while (!setStateChannel.isEmpty || !withStateChannel.isEmpty) {
+            var reducer = setStateChannel.poll()
+            while (reducer != null) {
+                val newState = state.reducer()
+                stateChannel.offer(newState)
+                state = newState
+                reducer = setStateChannel.poll()
+            }
+
+            withStateChannel.poll()?.let { withStateBlock ->
+                withStateBlock(state)
+            }
+        }
```

Comment:
the whole thing with prioritization setState over withState can be done via `select` instead of two queues + flush queue or priority queue as were suggested before. 
With `select` no flushQueue necessary, and get/set funs will be just offering to withStateChannel / setStateChannel. 

It could be done after removing FORCE_SYNCHRONOUS_STATE_STORES from this store though.